tools/ot/src/main/java/dk/alexandra/fresco/tools/ot/otextension/BristolOtReceiver.java

```diff
@@ -0,0 +1,134 @@
+    this.receiver = rotReceiver;
+    this.resources = resources;
+    this.network = network;
+    this.batchSize = batchSize;
+  }
+
+  /**
+   * Receive the serialized message from the current 1-out-of-2 OT.
+   *
+   * @param choiceBit
+   *          Choice-bit. False for message 0, true for message 1.
+   * @return The serialized message from the OT
+   */
+  public byte[] receive(Boolean choiceBit) {
+    // Check if there is still an unused random OT stored, if not, execute a
+    // random OT extension
+    if (offset < 0 || offset >= batchSize) {
+      choices = new StrictBitVector(batchSize, resources.getRandomGenerator());
+      randomMessages = receiver.extend(choices);
+      offset = 0;
+    }
+    // Notify the sender if it should switch the 0 and 1 messages around (s.t.
+    // the random choice bit in the preprocessed random OTs matches the true
+    // choice bit
+    sendSwitchBit(choiceBit);
+    // Receive the serialized adjusted messages
+    byte[] zeroAdjustment = network.receive(resources.getOtherId());
+    byte[] oneAdjustment = network.receive(resources.getOtherId());
+    byte[] res = doActualReceive(zeroAdjustment, oneAdjustment);
+    offset++;
+    return res;
+  }
+
+  /**
+   * Adjust the random, preprocessed message, to fit the specific message sent
+   * by the sender.
+   *
+   * @param zeroAdjustment
+   *          The adjustment value for the zero message
+   * @param oneAdjustment
+   *          The adjustment value for the one message
+   * @return The actual message
+   */
+  private byte[] doActualReceive(byte[] zeroAdjustment, byte[] oneAdjustment) {
+    if (zeroAdjustment.length != oneAdjustment.length) {
+      throw new MaliciousException(
+          "Sender gave adjustment messages of different length.");
+    }
+    byte[] adjustment;
+    if (choices.getBit(offset, false) == false) {
+      adjustment = zeroAdjustment;
+    } else {
+      adjustment = oneAdjustment;
+    }
+    adjustMessage(adjustment);
+    return adjustment;
+  }
+
+  /**
+   * Compute and send a bit indicating whether the sender should switch the zero
+   * and one message around.
+   *
+   * @param choiceBit
+   *          The actual choice bit of the receiver
+   */
+  private void sendSwitchBit(Boolean choiceBit) {
+    // Since we can only send a byte array we use 0x00 to indicate a 0-choice
+    // and 0x01 to indicate a 1-choice
+    byte[] switchBit = new byte[] { 0x00 };
+    // Set the choice to 0x01 if the sender should switch the 0 and 1 messages
+    if (choiceBit ^ choices.getBit(offset, false) == true) {
+      switchBit[0] = 0x01;
+    }
+    network.send(resources.getOtherId(), switchBit);
+  }
+
+  private void adjustMessage(byte[] adjustment) {
+    // Retrieve the random preprocessed message
+    byte[] randomMessage = randomMessages.get(offset).toByteArray();
+    // Use the random message as the seed to a PRG
+    Drbg currentPrg = new AesCtrDrbg(randomMessage);
```

Comment:
See comment on optimization on the Sender side.

Reply:
See comment on Sender side.